man/man5/rlm_chap.5
.TH rlm_chap 5 "3 February 2004" "" "FreeRADIUS Module"
.SH NAME
rlm_chap \- FreeRADIUS Module
.SH DESCRIPTION
The \fIrlm_chap\fP module provides CHAP functionality.
.PP
This module validates a user with CHAP authentication.
If called in Authorize, it will look for CHAP-Password
attribute in the Access-Request and adds an Auth-Type
attribute set to CHAP if the Config-Items list unless 
Auth-Type has already set.
.PP
CHAP authentication requires access to the clear-text
password for the user.  Standard Unix system authentication
or passwords encrypted via crypt() are not compatible
with CHAP.
.PP
.SH SECTIONS
.BR authorization,
.BR authentication
.PP
.SH FILES
.I /etc/raddb/radiusd.conf
.PP
.SH "SEE ALSO"
.BR radiusd (8),
.BR radiusd.conf (5)
.SH AUTHOR
Chris Parker, cparker@segv.org



doc/ChangeLog
FreeRADIUS 2.2.10 Wednesday 30 Sep 2015 17:00:00 EDT, urgency=medium
	Feature improvements
	* None.

	Bug fixes
	* Work around OpenSSL 1.0.2 problems, which cause failures
	  in TLS-based EAP methods.
	* Revert RedHat contributed bug which removes run-time checks
	  for OpenSSL consistency.
	* Allow OCSP responder URL to be later in the packet.
	  Fix by Ean Pasternak.
	* Catch empty subject and non-existent issuer cert in OCSP.
	  Fix by Ean Pasternak.
	* Allow non-FIPS for MD5.
	  Fix by Ean Pasternak.

FreeRADIUS 2.2.9 Wednesday 30 Sep 2015 17:00:00 EDT, urgency=medium
	Feature improvements
	* None.

	Bug fixes
	* Fix Tunnel-Password crash from home server.
	  Found by Denis Andzakovic.
	* Fix timer issue when proxying.
	* Update EAP-TTLS so that MPPE keys are correctly
	  calculated with TLSv1.2.
	* Always delete MS-MPPE-* from the TTLS inner tunnel.
	  This allows TTLS / EAP-MSCHAPv2 to work.
	* Don't fall through in empty "case" statements.
	  Fixes #1274.

FreeRADIUS 2.2.8 Thursday 09 Jul 2015 11:00:00 EDT, urgency=medium
	Feature improvements
	* None.

	Bug fixes
	* Fixes for clients tied to virtual servers.  If there is
	  no "listen" section there, clients use the main "listen"
	  section.
	* Remove compiler warnings
	* Print out correct filenames in debug mode
	* Allow post-auth section to return "reject".  This turns
	  the response into Access-Reject.
	* Set X509_V_FLAG_CRL_CHECK_ALL, and add "check_all_crl"
	  to eap.conf.  Fixes oCert CVE-2015-4680.

FreeRADIUS 2.2.7 Wednesday 22 Apr 2015 14:00:00 EST, urgency=medium
	Feature improvements
	* Allow "eap" to be listed in Post-Auth-Type Reject
	  so that it sends EAP-Fail and Message-Authenticator.

	Bug fixes
	* Fix typo in code checking for blocked threads.
	  Closes #880.
	* Added more $(EXEEXT) to module utilities so that it
	  builds on Cygwin.  Closes #875.
	* Note that we don't need to generate ephemeral RSA keys.
	* Port detail file fixes from v3.
	* Use correct destination port for replies to DHCP relay.
	* rlm_perl can store multiple tagged attributes of the
	  same name.
	* Update EAP-TLS methods for TLSv1.2
	* Fix load-balance sections.  Closes #945

FreeRADIUS 2.2.6 Tuesday 18 Nov 2014 15:00:00 EST, urgency=medium
	Feature improvements
	* When supported by OpenSSL, allow TLS 1.1 and TLS 1.2
	  in EAP methods.

	Bug fixes
	* Fix redundant-load-balance blocks to try other modules in
	  the group if one fails.
	* Fix potential read into uninitialised memory in rlm_pap
	  when normalising octet type attributes containing password
	  hashes. This is very unlikely to happen in the wild.
	* Don't stop decoding DHCP options if we find a padding
	  option.
	* Define sig_t on systems which don't have it.  Closes #765
	* When clients are loaded from SQL, allow them to be tied
	  to a virtual server.
	* Prevent race conditions between fork and wait for child.
	  Patch from James Rouzier.
	* Allow UTF-8 characters in SQL.
	* Back-port udpfromto fixes from v3

FreeRADIUS 2.2.5 Monday 28 Apr 2014 15:20:00 EDT, urgency=medium
	Feature improvements
	* Update dictionary.terena.
	* expose server version via %v.  Patch from Alan Buxey.
	* Forbid running with vulnerable versions of OpenSSL.
	  See "allow_vulnerable_openssl" in the "security"
	  subsection of "radiusd.conf"
	* Catch underlying "heartbleed" problem, so that nothing bad
	  happens even when using a vulnerable version of OpenSSL.
	* Add dictionary.zte

	Bug fixes
	* Minor changes to build on Sun.
	* Print non-ASCII characters as octal in linelog.  Closes #578
	* close stdout in daemon mode.
	* Fix zombie period calculation.  Closes #579

FreeRADIUS 2.2.4 Wednesday 19 Mar 2014 13:20:00 EDT, urgency=medium
	Feature improvements
	* A "panic_action" can be set to have the server dump a gdb
	  log on SEGV or other fatal error.
	* allow radmin command "set module status <module> <code>"
	  which can be used to forcibly enable/disable modules.

	Bug fixes
	* If the server fails to bind() after fork(), that is now
	  reported to the parent, which exits with an error.
	* Session / delay times in MySQL are unsigned int.
	* Use --tag=CC for libtool.  Closes #497.  Because libtool
	  is too stupid to notice that compiling means compilation.
	* Fix bug when copying attributes for vendors > 32767
	* Fix behaviour on FreeBSD where sending packets from an interface
	  bound to an IP address would fail when the server was built with
	  udpfromto.
	* Don't fail config check if were listening on an IP which is
	  also a home server. Some deployments have valid reasons
	  to loop packets back to another virtual server.
	* Use correct port when DHCP relaying.
	* Set source IP address for DHCP packets from DHCP-Server-IP-Address,
	  or DHCP-DHCP-Server-Identifier, if we're unable to otherwise
	  determine the source IP.

FreeRADIUS 2.2.3 Wednesday 11 Dec 2013 15:00:00 EST, urgency=medium
	Feature improvements
	* Added dictionary.efficientip, dictionary.alcatel-lucent-aaa
	* Allow zero length DN strings in rlm_ldap.
	* If Password-With-Header has no header, assume it is
	  Cleartext-Password.

	Bug fixes
	* Make the server build when DHCP is enabled
	* Don't crash if there's no Post-Proxy-Type Reject.
	* Use correct fields for X509 attributes in certificates
	* Install threads.h making it possible to link against the
	  installed headers again.
	* Initialize SSL once in "main", instead of rlm_eap_tls.
	  Some client libraries may need SSL.

FreeRADIUS 2.2.2 Wednesday 30 Oct 2013 9:30:00 DST, urgency=medium
	Feature improvements
	* Add "timeout" to exec, and "ntlm_auth_timeout" to mschap.
	  So that run-away child processes are caught earlier.
	* Print out thread number for "unresponsive child".

	Bug fixes
	* Fix erroneous fall-through in "case" statements
	* Fix priority handling in new module handling code
	* Fix threading issue with Perl.  Closes #436
	* Fix EAP-TLS check_cert_issuer when X509v2 extensions
	  existed.  Patch from David Wood.
	* Fix pointer references in rlm_python.

src/modules/rlm_chap/rlm_chap.c
/*
 *   This program is is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation; either version 2 of the License, or (at
 *   your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program; if not, write to the Free Software
 *   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
 */

/**
 * $Id$
 * @file rlm_chap.c
 * @brief Process chap authentication requests.
 *
 * @copyright 2001,2006  The FreeRADIUS server project
 * @copyright 2001  Kostas Kalevras <kkalev@noc.ntua.gr>
 */
RCSID("$Id$")

#include <freeradius-devel/radiusd.h>
#include <freeradius-devel/modules.h>

static rlm_rcode_t CC_HINT(nonnull) mod_authorize(UNUSED void *instance, REQUEST *request)
{
	if (!fr_pair_find_by_num(request->packet->vps, 0, PW_CHAP_PASSWORD, TAG_ANY)) {
		return RLM_MODULE_NOOP;
	}

	if (fr_pair_find_by_num(request->control, 0, PW_AUTH_TYPE, TAG_ANY) != NULL) {
		RWDEBUG2("&control:Auth-Type already set.  Not setting to CHAP");
		return RLM_MODULE_NOOP;
	}

	RINDENT();
	RDEBUG("&control:Auth-Type := CHAP");
	REXDENT();
	pair_make_config("Auth-Type", "CHAP", T_OP_EQ);

	return RLM_MODULE_OK;
}


/*
 *	Find the named user in this modules database.  Create the set
 *	of attribute-value pairs to check and reply with for this user
 *	from the database. The authentication code only needs to check
 *	the password, the rest is done here.
 */
static rlm_rcode_t CC_HINT(nonnull) mod_authenticate(UNUSED void *instance,
				     REQUEST *request)
{
	VALUE_PAIR *password, *chap;
	uint8_t pass_str[FR_MAX_STRING_LEN];

	if (!request->username) {
		REDEBUG("&request:User-Name attribute is required for authentication");
		return RLM_MODULE_INVALID;
	}

	chap = fr_pair_find_by_num(request->packet->vps, 0, PW_CHAP_PASSWORD, TAG_ANY);
	if (!chap) {
		REDEBUG("You set '&control:Auth-Type = CHAP' for a request that "
			"does not contain a CHAP-Password attribute!");
		return RLM_MODULE_INVALID;
	}

	if (chap->vp_length == 0) {
		REDEBUG("&request:CHAP-Password is empty");
		return RLM_MODULE_INVALID;
	}

	if (chap->vp_length != CHAP_VALUE_LENGTH + 1) {
		REDEBUG("&request:CHAP-Password has invalid length");
		return RLM_MODULE_INVALID;
	}

	password = fr_pair_find_by_num(request->control, 0, PW_CLEARTEXT_PASSWORD, TAG_ANY);
	if (password == NULL) {
		if (fr_pair_find_by_num(request->control, 0, PW_USER_PASSWORD, TAG_ANY) != NULL){
			REDEBUG("!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!");
			REDEBUG("!!! Please update your configuration so that the \"known !!!");
			REDEBUG("!!! good\" cleartext password is in Cleartext-Password,  !!!");
			REDEBUG("!!! and NOT in User-Password.                            !!!");
			REDEBUG("!!!						          !!!");
			REDEBUG("!!! Authentication will fail because of this.	          !!!");
			REDEBUG("!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!");
		}

		REDEBUG("&control:Cleartext-Password is required for authentication");
		return RLM_MODULE_FAIL;
	}

	fr_radius_encode_chap_password(pass_str, request->packet, chap->vp_octets[0], password);

	if (RDEBUG_ENABLED3) {
		uint8_t const *p;
		size_t length;
		VALUE_PAIR *vp;
		char buffer[CHAP_VALUE_LENGTH * 2 + 1];

		RDEBUG3("Comparing with \"known good\" &control:Cleartext-Password value \"%s\"",
			password->vp_strvalue);

		vp = fr_pair_find_by_num(request->packet->vps, 0, PW_CHAP_CHALLENGE, TAG_ANY);
		if (vp) {
			RDEBUG2("Using challenge from &request:CHAP-Challenge");
			p = vp->vp_octets;
			length = vp->vp_length;
		} else {
			RDEBUG2("Using challenge from authenticator field");
			p = request->packet->vector;
			length = sizeof(request->packet->vector);
		}

		fr_bin2hex(buffer, p, length);
		RINDENT();
		RDEBUG3("CHAP challenge : %s", buffer);

		fr_bin2hex(buffer, chap->vp_octets + 1, CHAP_VALUE_LENGTH);
		RDEBUG3("Client sent    : %s", buffer);

		fr_bin2hex(buffer, pass_str + 1, CHAP_VALUE_LENGTH);
		RDEBUG3("We calculated  : %s", buffer);
		REXDENT();
	} else {
		RDEBUG2("Comparing with \"known good\" Cleartext-Password");
	}

	if (fr_radius_digest_cmp(pass_str + 1, chap->vp_octets + 1, CHAP_VALUE_LENGTH) != 0) {
		REDEBUG("Password comparison failed: password is incorrect");
		return RLM_MODULE_REJECT;
	}

	RDEBUG("CHAP user \"%s\" authenticated successfully", request->username->vp_strvalue);

	return RLM_MODULE_OK;
}


/*
 *	Access-Requests can have the CHAP-Challenge implicitly taken
 *	from the request authenticator.  If the NAS has done that,
 *	then we need to copy the data to a real CHAP-Challenge
 *	attribute when proxying.  Otherwise when we proxy the request,
 *	the new authenticator is different, and the CHAP calculations
 *	will fail.
 */
static rlm_rcode_t CC_HINT(nonnull) mod_pre_proxy(UNUSED void *instance,
				 REQUEST *request)
{
	VALUE_PAIR *vp;

	/*
	 *	For Access-Requests, which have CHAP-Password,
	 *	and no CHAP-Challenge, copy it over from the request.
	 */
	if (request->packet->code != PW_CODE_ACCESS_REQUEST) return RLM_MODULE_NOOP;

	if (!fr_pair_find_by_num(request->proxy->packet->vps, 0, PW_CHAP_PASSWORD, TAG_ANY)) return RLM_MODULE_NOOP;

	if (fr_pair_find_by_num(request->proxy->packet->vps, 0, PW_CHAP_CHALLENGE, TAG_ANY)) return RLM_MODULE_NOOP;

	vp = radius_pair_create(request->proxy->packet, &request->proxy->packet->vps, PW_CHAP_CHALLENGE, 0);
	if (!vp) return RLM_MODULE_FAIL;

	fr_pair_value_memcpy(vp, request->packet->vector, sizeof(request->packet->vector));

	return RLM_MODULE_OK;
}


/*
 *	The module name should be the only globally exported symbol.
 *	That is, everything else should be 'static'.
 *
 *	If the module needs to temporarily modify it's instantiation
 *	data, the type should be changed to RLM_TYPE_THREAD_UNSAFE.
 *	The server will then take care of ensuring that the module
 *	is single-threaded.
 */
extern rad_module_t rlm_chap;
rad_module_t rlm_chap = {
	.magic		= RLM_MODULE_INIT,
	.name		= "chap",
	.methods = {
		[MOD_AUTHENTICATE]	= mod_authenticate,
		[MOD_AUTHORIZE]		= mod_authorize,
		[MOD_PRE_PROXY]		= mod_pre_proxy
	},
};
